sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +203,217 @@ object Hash {
+    }
+  }
+
+  /** Class with additional methods to hash nodes. Uses a single MessageDigest to hash the entire node including all its values.
+    */
+  private sealed abstract class NodeBuilder(
+      version: NodeHashVersion,
+      hashTracer: HashTracer,
+  ) extends TransactionBuilder(hashTracer) {
+
+    override def addVersion: this.type = {
+      super.addVersion
+        .addByte(version.id, s"${formatByteToHexString(version.id)} (Node Encoding Version)")
+    }
+
+    def addNode(node: Node, nodes: Map[NodeId, Node]): NodeBuilder
+  }
+
+  private object NodeBuilder {
+    private[lf] val LanguageVersionToNodeBuilder: Map[LanguageVersion, HashTracer => NodeBuilder] =
+      Map(
+        LanguageVersion.v2_1 -> (hashTracer => new NodeBuilderV1(hashTracer))
+      )
+
+    private[crypto] sealed abstract class NodeTag(val tag: Byte)
+    private[crypto] object NodeTag {
+      case object CreateTag extends NodeTag(0)
+      case object ExerciseTag extends NodeTag(1)
+      case object FetchTag extends NodeTag(2)
+      case object LookupTag extends NodeTag(3)
+      case object RollbackTag extends NodeTag(4)
+    }
+
+    @throws[NodeHashingError]
+    private[lf] def builderForVersion(
+        version: LanguageVersion,
+        hashTracer: HashTracer,
+    ): NodeBuilder =
+      NodeBuilder.LanguageVersionToNodeBuilder
+        .getOrElse(version, throw NodeHashingError.UnsupportedLanguageVersion(version))
+        .apply(hashTracer)
+
+    private[lf] def builderForNode(
+        node: Node,
+        hashTracer: HashTracer,
+    ): NodeBuilder = node.optVersion match {
+      case Some(version) => builderForVersion(version, hashTracer)
+      // If the node is version agnostic, the builder doesn't matter so pick V1
+      // This only applies to rollback nodes
+      case None => new NodeBuilderV1(hashTracer)
+    }
+  }
+
+  private final class NodeBuilderV1(hashTracer: HashTracer)
+      extends NodeBuilder(NodeHashVersion.V1, hashTracer) {
+    private def addCreateNode(create: Node.Create): this.type = {
+      addContext("Create Node")
+        .addByte(NodeBuilder.NodeTag.CreateTag.tag, "Node Tag")
+        .withContext("Node Version")(_.addString(TransactionVersion.toProtoValue(create.version)))
+        .withContext("Contract Id")(_.addCid(create.coid))
+        .withContext("Package Name")(_.addString(create.packageName))
+        .withContext("Template Id")(_.addIdentifier(create.templateId))
+        .withContext("Arg")(_.addTypedValue(create.arg))
+        .withContext("Signatories")(_.addStringSet(create.signatories))
+        .withContext("Stakeholders")(_.addStringSet(create.stakeholders))
+    }
+
+    private def addFetchNode(fetch: Node.Fetch): NodeBuilder = {
+      addContext("Fetch Node")
+        .addByte(NodeBuilder.NodeTag.FetchTag.tag, "Node Tag")
+        .withContext("Node Version")(_.addString(TransactionVersion.toProtoValue(fetch.version)))
```

Comment:
I think it is better to put the version first, since the type of node are depending on that. 
More a gut feeling.

```suggestion
        .withContext("Node Version")(_.addString(TransactionVersion.toProtoValue(fetch.version)))
        .addByte(NodeBuilder.NodeTag.FetchTag.tag, "Node Tag")
```